### PR TITLE
add restriction on created space type

### PR DESCRIPTION
**Pull request by asf-algoo**
related to #3602

## Checkpoints

**For developers**

- [x] If relevant, manual tests have been done to ensure the stability of the whole application and that the involved feature works
- [x] The original issue is up to date w.r.t the latest discussions and contains a short summary of the implemented solution
- [x] Automated tests covering the feature or the fix, have been written, deemed irrelevant (give the reason), or an issue has been created to implement the test (give the link)

**For code reviewers**

- [x] The code is clear enough
- [x] If there are FIXMEs in the code, related issues are mentioned in the FIXME
- [x] If there are TODOs, NOTEs or HACKs in code, the date and the developer initials are present

**For testers**

- [x] Manual, quality tests have been done

### CODE REVIEW

File: backend/development.ini.sample [hunk 1]
```diff
@@ -101,6 +101,14 @@ basic_setup.session_secret = change_this_value_please!
 # be loaded, allowing to add new feature to an existing Tracim.
 ; plugin.folder_path = %(here)s/plugins
 
+### Workspace ###
+# list of access_type allowed for new workspace. Access_type available in Tracim are:
+# - confidential: not visible publicly, access only by invitation workspace_manager/admin
+# - on_request: publicly visible, access by invitation or request with validation by workspace manager/admin)
+# - open: publicly visible and access by invitation or request without validation
+; workspace.allowed_access_types = confidential,on_request,open
```
Comment (grignards) [inline backend/development.ini.sample:109]: Maybe add this is the default value?

File: backend/tracim_backend/exceptions.py [hunk 2]
```diff
@@ -319,6 +319,10 @@ class EmptyLabelNotAllowed(EmptyValueNotAllowed):
     pass
 
 
+class UnallowedWorkspaceAccessType(TracimException):
```
Comment (grignards) [inline backend/tracim_backend/exceptions.py:322]:
```suggestion
class DisallowedWorkspaceAccessType(TracimException):
```

File: backend/tracim_backend/lib/core/workspace.py [hunk 3]
```diff
@@ -11,6 +11,7 @@
 from tracim_backend.config import CFG
 from tracim_backend.exceptions import AgendaServerConnectionError
 from tracim_backend.exceptions import EmptyLabelNotAllowed
+from tracim_backend.exceptions import UnallowedWorkspaceAccessType
```
Comment (grignards) [inline backend/tracim_backend/lib/core/workspace.py:14]:
```suggestion
from tracim_backend.exceptions import DisallowedWorkspaceAccessType
```

File: backend/tracim_backend/lib/core/workspace.py [hunk 4]
```diff
@@ -107,7 +108,10 @@ def create_workspace(
             raise UserNotAllowedToCreateMoreWorkspace("User not allowed to create more workspace")
         if not label:
             raise EmptyLabelNotAllowed("Workspace label cannot be empty")
-
+        if access_type not in self._config.WORKSPACE__ALLOWED_ACCESS_TYPES:
+            raise UnallowedWorkspaceAccessType(
```
Comment (grignards) [inline backend/tracim_backend/lib/core/workspace.py:112]:
```suggestion
            raise DisallowedWorkspaceAccessType(
```

File: backend/tracim_backend/views/core_api/workspace_controller.py [hunk 5]
```diff
@@ -16,6 +16,7 @@
 from tracim_backend.exceptions import ParentNotFound
 from tracim_backend.exceptions import RoleAlreadyExistError
 from tracim_backend.exceptions import UnallowedSubContent
+from tracim_backend.exceptions import UnallowedWorkspaceAccessType
```
Comment (grignards) [inline backend/tracim_backend/views/core_api/workspace_controller.py:19]:
```suggestion
from tracim_backend.exceptions import DisallowedWorkspaceAccessType
```

File: backend/tracim_backend/views/core_api/workspace_controller.py [hunk 6]
```diff
@@ -145,6 +146,7 @@ def workspaces(self, context, request: TracimRequest, hapic_data=None):
 
     @hapic.with_api_doc(tags=[SWAGGER_TAG__WORKSPACE_ENDPOINTS])
     @hapic.handle_exception(EmptyLabelNotAllowed, HTTPStatus.BAD_REQUEST)
+    @hapic.handle_exception(UnallowedWorkspaceAccessType, HTTPStatus.BAD_REQUEST)
```
Comment (grignards) [inline backend/tracim_backend/views/core_api/workspace_controller.py:149]:
```suggestion
    @hapic.handle_exception(DisallowedWorkspaceAccessType, HTTPStatus.BAD_REQUEST)
```

File: backend/tracim_backend/views/core_api/workspace_controller.py [hunk 7]
```diff
@@ -172,6 +174,7 @@ def update_workspace(self, context, request: TracimRequest, hapic_data=None):
...
```
Comment (grignards) [inline backend/tracim_backend/views/core_api/workspace_controller.py:177]:
```suggestion
    @hapic.handle_exception(DisallowedWorkspaceAccessType, HTTPStatus.BAD_REQUEST)
```

**Review by grignards — APPROVED** [review]

File: backend/development.ini.sample (hunk 1 above)
Comment (inkhey) [reply backend/development.ini.sample:109]: default value are all showed this way with the ";" it's explain at the beginning of the config file.

**Review by inkhey — CHANGES_REQUESTED** [review]
need updating config parameters list

File: backend/development.ini.sample (hunk 1 above)
Comment (inkhey) [reply backend/development.ini.sample:109]: Config file begin with:
```ini
## File Syntax
# Comments start by `;` or `#`.
# In the sample config file:
#  - `;` commented lines are used to indicate the default values used by Tracim;
#  - `#` commented lines are used for comments/examples.
# Boolean values can be `True` or `False`
# All parameters starting with `basic_setup` can be used as variables in other parameters.
# For simple needs they are all you need to customize.

```
With this documentation, Is it ok for you ?

File: backend/tracim_backend/exceptions.py (hunk 2 above)
Comment (inkhey) [reply backend/tracim_backend/exceptions.py]: Renamed everywhere, Change the Error name too.

**Review by inkhey — APPROVED** [review]

**Comment by inkhey** [comment]
To test this:
- update both backend and frontend in this branch.
- play with `workspace.allowed_access_types`, removing confidential from the list should in current frontend implementation do not allow to create space anymore, but should not disallow the capability to update an existing workspace.
- setting an invalid access_type name should not allow to start tracim.

File: backend/development.ini.sample (hunk 1 above)
Comment (raphj) [inline backend/development.ini.sample:109]:
```suggestion
# list of access_type allowed for new workspaces. In Tracim; these values for access_type are available:
# - confidential: not visible publicly, access only by invitation workspace_manager/admin
# - on_request: publicly visible, access by invitation or request with validation by workspace manager/admin)
# - open: publicly visible and access by invitation or request without validation
; workspace.allowed_access_types = confidential,on_request,open
```

File: backend/tracim_backend/lib/core/workspace.py [hunk 8]
```diff
@@ -107,7 +108,10 @@ def create_workspace(
             raise UserNotAllowedToCreateMoreWorkspace("User not allowed to create more workspace")
         if not label:
             raise EmptyLabelNotAllowed("Workspace label cannot be empty")
-
+        if access_type not in self._config.WORKSPACE__ALLOWED_ACCESS_TYPES:
+            raise DisallowedWorkspaceAccessType(
+                'Access type "{}" for workspace is not allowed'.format(access_type.name)
```
Comment (raphj) [inline backend/tracim_backend/lib/core/workspace.py:113]:
```suggestion
                'Access type "{}" is not allowed for this workspace'.format(access_type.name)
```

File: backend/tracim_backend/tests/functional/test_system.py [hunk 9]
```diff
@@ -294,3 +294,30 @@ def test_api__get_error_code_endpoint__ok__200__nominal_case(self, web_testapp):
...
+
+@pytest.mark.usefixtures("base_fixture")
+class TestWorkspaceAccessType(object):
+    """
+    Tests for /api/system/workspace_access_types
+    """
+
+    def test_api__get_workspace_access_type__ok_200__nominal_case(self, web_testapp):
+        """
+        Get alls allowed workspace access types list with a registered user.
```
Comment (raphj) [inline backend/tracim_backend/tests/functional/test_system.py:307]:
```suggestion
        Get the list of allowed workspace access types with a registered user.
```

**Review by raphj — APPROVED** [review]

File: backend/development.ini.sample (hunk 1 above)
Comment (grignards) [reply backend/development.ini.sample:109]: Mostly, I just now just find strange to put the default commented value before the example, I would have put the example before.

Comment (inkhey) [reply backend/development.ini.sample:109]: I don't have much opinion on this subject, we do in this order at multiple place in config file without any trouble until now.
I suggest to get opinion from both @PhilippeAccorsi and @raphj and change the order according to that if needed in another issue .

Comment (grignards) [reply backend/development.ini.sample:109]: Yes, this is also why I approved the changes: my comment is not blocking.

Comment (raphj) [reply backend/development.ini.sample:109]: I agree with @grignards, we should put the example before the default value.

Comment (PhilippeAccorsi) [reply backend/development.ini.sample:109]: If I remember correctly we have done a big modification in development.ini file in Tracim 2.7.x with new naming process and validate this choice (with some others dev). Modification in this file is not simple on existing installation so for me its important to define structure of file for a long time period.

I think its not necessary to change structure of this file each years ;)

So I think we can create an issue about wording of this file, and try to find what is the better solution (what is done on other projet, what we really want, etc.)

**Comment by grignards** [comment]
Tested on branch.

**Pull request merged**